metronome/storage/src/io/iohk/metronome/storage/KVCollection.scala

```diff
@@ -0,0 +1,29 @@
+package io.iohk.metronome.storage
+
+import scodec.Codec
+
+/** Storage for a specific type of data, e.g. blocks, in a given namespace.
+  *
+  * We should be able to string together KVStore operations across multiple
+  * collections and execute them in one batch.
+  */
+class KVCollection[N, K: Codec, V: Codec](namespace: N) {
```

Comment:
Good question! The reason `N` doesn't require a codec is precisely because it _has_ to be a type that is common across all `KVCollection` instances so that they can be composed together. To put it differently: it doesn't need to be captured on the collection level because it is knowledge available to the compiler, otherwise it cannot execute the program. By contract, the `K` and `V` are specific to each collection, so their codecs can be different by each `Get`, `Put` and `Delete`, depending on which namespace they are accessing. In theory the codecs could be registered in a `Map` somewhere, but this way they can provide more type safety.

The return value of a `get` for example is `KVStore[N, Option[V]]`, so the `N` is preserved there, so whatever it is, the user will have to pass it to a compiler that takes the same type. I wanted this to be as generic as possible, and fixing to `IndexedSeq[Byte]` seemed an implementation detail of the current RocksDB implementation, while `Codec` was - well I thought if I wanted to factor _that_ out then it would have become needlessly complicated.

Reply:
oh I think I understand the dilema here. 

So my personal choice would be to fix it to `IndexedSeq[Byte]` as it will be fixed to this either way when used with rocks db, but if you think there is some value in having it generic then we can leave it that way. (we can always revisit this decision in the future)

Reply:
Yeah, on one hand it would make the lambdas go away, on the other hand I think what we're doing [here](https://github.com/input-output-hk/mantis/blob/develop/src/main/scala/io/iohk/ethereum/db/storage/Namespaces.scala#L4-L14) is suspect: do we really have to have single character namespaces? They all [turn into byte arrays](https://github.com/input-output-hk/mantis/blob/develop/src/main/scala/io/iohk/ethereum/db/dataSource/RocksDbDataSource.scala#L306) explicitly, so we might use something more readable, like a `String` or whatever. At least I'm not convinced it's something inherent to the domain that namespaces have to be obscure bytes.

Reply:
Another possible design alternative would be to make the namespace carry around the codecs and the key-value types, something like this:

```scala
trait Namespace {
  type Key
  type Value
}

trait RocksDBNamespace extends Namespace {
  val namespace: IndexedSeq[Byte]
  val keyCodec: Codec[Key]
  val valueCodec: Codec[Value]
}
```

And change the `KVStoreOps` to look like this:
```scala
  case class Put[N](namespace: N, key: N#Key, value: N#Value) extends KVStoreWriteOp[N, Unit]
```
(not sure if this compiles).

So that only the RocksDB stuff has to know about codecs.